modules/withRouter.js

```diff
@@ -1,23 +1,40 @@
 import React from 'react'
 import hoistStatics from 'hoist-non-react-statics'
 import { routerShape } from './PropTypes'
+import warning from './routerWarning'
 
 function getDisplayName(WrappedComponent) {
   return WrappedComponent.displayName || WrappedComponent.name || 'Component'
 }
+export default function withRouter(options) {
+  const wrapWithRouter = function (WrappedComponent) {
+    const WithRouter = React.createClass({
+      contextTypes: { router: routerShape },
+      propTypes: { router: routerShape },
+      getWrappedInstance() {
+        warning(options && options.withRef, 'To access the wrappedInstance you must provide {withRef : true} as the first argument of the withRouter call')
+        return this.refs.wrappedInstance
+      },
+      render() {
+        const router = this.props.router || this.context.router
+        if(options && options.withRef) {
+          return <WrappedComponent {...this.props} ref="wrappedInstance" router={router} />
+        }else{
+          return <WrappedComponent {...this.props} router={router} />
+        }
+      }
+    })
+    WithRouter.displayName = `withRouter(${getDisplayName(WrappedComponent)})`
+    WithRouter.WrappedComponent = WrappedComponent
 
-export default function withRouter(WrappedComponent) {
-  const WithRouter = React.createClass({
-    contextTypes: { router: routerShape },
-    propTypes: { router: routerShape },
-    render() {
-      const router = this.props.router || this.context.router
-      return <WrappedComponent {...this.props} router={router} />
-    }
-  })
+    return hoistStatics(WithRouter, WrappedComponent)
+  }
 
-  WithRouter.displayName = `withRouter(${getDisplayName(WrappedComponent)})`
-  WithRouter.WrappedComponent = WrappedComponent
-
-  return hoistStatics(WithRouter, WrappedComponent)
+  if(typeof(options) === 'function') {
```

Comment:
Why? Just make `options` be the second arg. This is the less common use case and it's not worth the deprecation or making ergonomics worse for the common case.